### PR TITLE
[improve][java-client] Only trigger the batch receive timeout when having pending batch receives requests

### PR DESCRIPTION
**Pull request by codelipenghui**
### Motivation

The consumer will apply the default batch receive policy even if the user will not use the batch receive API.

https://github.com/apache/pulsar/blob/6704f12104219611164aa2bb5bbdfc929613f1bf/pulsar-client-api/src/main/java/org/apache/pulsar/client/api/BatchReceivePolicy.java#L60-L61

This will consume lots of CPU if the client has many consumers (100k consumers)

image

[consumer-cpu-threads.html.txt](https://github.com/apache/pulsar/files/8948238/consumer-cpu-threads.html.txt)

The Pulsar perf tool can also reproduce the problem if run the test with many consumers

### Modification

If there is no pending batch receive operation for a consumer, no need to trigger the
batch timeout task periodically. We can only start the timeout check after adding batch
receive request to pending request queue.

Remove the lock in MultiTopicsConsumerImpl as #10352 does

### Verification

Added new test to verify the batch receive timeout task will not start if no batch
receive request

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

**Review by merlimat — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java [hunk 1]
```diff
@@ -164,12 +164,14 @@ protected ConsumerBase(PulsarClientImpl client, String topic, ConsumerConfigurat
             this.batchReceivePolicy = BatchReceivePolicy.DEFAULT_POLICY;
         }
 
+        initReceiverQueueSize();
+    }
+
+    protected void triggerBatchReceiveTimeoutTask() {
```
Comment (michaeljmarshall) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:170]: If I am reading this correctly, this method will get called with every call to `batchReceive` or `batchReceiveAsync`. That means we could end up with several concurrent timer tasks for a single consumer. The current solution only has one task at a time per consumer. Given that the `doPendingBatchReceiveTask` method schedules the next task based on the `diff` to the next `opBatchReceive` timeout, I think this method should only schedule a timer task when `batchReceiveTimeout` is `null`.

**Review by michaeljmarshall — CHANGES_REQUESTED** [review]
@codelipenghui - this is a great improvement. I have one concern about concurrent timer tasks, but otherwise, this looks good.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:170]: > If I am reading this correctly, this method will get called with every call to batchReceive or batchReceiveAsync.

No, we only call this method if no messages in the receiver queue. 

> I think this method should only schedule a timer task when batchReceiveTimeout is null.

Yes, make sense. I think I missed this part.

Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:170]: @michaeljmarshall Fixed.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java [hunk 2]
```diff
@@ -164,12 +164,14 @@ protected ConsumerBase(PulsarClientImpl client, String topic, ConsumerConfigurat
             this.batchReceivePolicy = BatchReceivePolicy.DEFAULT_POLICY;
         }
 
-        if (batchReceivePolicy.getTimeoutMs() > 0) {
+        initReceiverQueueSize();
+    }
+
+    protected void triggerBatchReceiveTimeoutTask() {
+        if (!hasBatchReceiveTimeout() && batchReceivePolicy.getTimeoutMs() > 0) {
```
Comment (Jason918) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:171]: Both `internalPinnedExecutor` and `client.timer()` threads changes `batchReceiveTimeout`, seems not thread safe.

Comment (codelipenghui) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:171]: The timer thread will not access the `batchReceiveTimeout`

https://github.com/apache/pulsar/pull/16160/files#diff-7a1c869c4bbefdfc72fcee6074b4dfeeec318334d48d998c1cff5e09713b0834R951-R953

But I found another place `closeConsumerTasks` will also access the `batchReceiveTimeout` but can happen on the IO thread, the user's thread, but only get the `batchReceiveTimeout`, will not change it.  And we have `volatile` for `batchReceiveTimeout`.

Comment (Jason918) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:171]: My bad, `doPendingBatchReceiveTask` runs in `internalPinnedExecutor`

**Review by Jason918 — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java (hunk 1 above)
Comment (michaeljmarshall) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java:170]: Thanks!

**Review by michaeljmarshall — APPROVED** [review]
LGTM

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java [hunk 3]
```diff
@@ -695,18 +688,15 @@ private ConsumerConfigurationData<T> getInternalConsumerConfig() {
 
     @Override
     public void redeliverUnacknowledgedMessages() {
-        lock.writeLock().lock();
-        try {
+        internalPinnedExecutor.execute(() -> {
             CONSUMER_EPOCH.incrementAndGet(this);
```
Comment (congbobo184) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java:692]: use `internalPinnedExecutor` then `CONSUMER_EPOCH.incrementAndGet(this);` the epoch will not work

**Pull request merged**